### PR TITLE
BIP 337: Compressed Transactions

### PR DESCRIPTION
**Pull request by TomBriar**
This document proposes a serialization scheme for compressing bitcoin transactions. The compressed bitcoin transactions can reach a serialized size of less than 50% of the original serialized transaction. One of the methods for compressing, involves reducing the transaction outpoints in a potentially lossy way, therefore it is an optional path for compression. Without compressing the outpoints, compressed transactions still reach less then 70% of the original size.

Reference Implementation:
https://github.com/bitcoin/bitcoin/pull/29134

The reference implementation includes two new RPC endpoints to demonstrate how to use compressed transactions. The compressrawtransaction takes in a raw hex-encoded transaction and returns a compressed hex-encoded transaction, while the decompressrawtransaction accepts a compressed hex-encoded transaction and returns the original raw hex-encoded transaction. 

Thanks-
Tom.

### CODE REVIEW

File: bip-tombriar-compressed-transactions.mediawiki [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+==== General Schema ====
+
+{| class="wikitable" style="margin:auto"
+|-
+! Name !! Width !! Description
+|-
+| Transaction metadata || 1 bytes || Information on the structure of the transaction. See [[#transaction-metadata|Transaction Metadata]]
+|-
+| Version || 0-5 Bytes || An optional CompactSize containing the transactions version.
```
Comment (apoelstra) [inline bip-tombriar-compressed-transactions.mediawiki:44]: grammar: `transactions` should be `transaction` or `transaction's` here and in the 3 lines below.

I'd also change "optional" to "if present according to the metadata field, a CompactSize encoding the transaction version". So it's clear on a first read what "optional" means.

File: bip-tombriar-compressed-transactions.mediawiki [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+|}
+
+<span id="input-metadata"></span>
+==== Input Metadata ====
+
+{| class="wikitable" style="margin:auto"
+|-
+! Name !! Width !! Description
+|-
+| Compressed Signature || 1 Bit || Signature compression flag. For P2TR: 1 for keyspend, 0 for scriptspend; For P2SH: 0 for p2sh, 1 for p2sh-wpkh.
```
Comment (apoelstra) [inline bip-tombriar-compressed-transactions.mediawiki:84]: Maybe mention that for optut types other than P2TR and P2SH this bit is unused.

File: bip-tombriar-compressed-transactions.mediawiki [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+|-
+| Standard Hash || 1 Bit || A flag to determine if this Input's Signature Hash Type is standard (0x00 for Taproot, 0x01 for Legacy/Segwit).
+|-
+| Standard Squence || 2 Bits || A CompactSize flag for the inputs sequence. Encode literal values as follows: 1 = 0x00000000, 2 = 0xFFFFFFFE, 3 = 0xFFFFFFFF.
```
Comment (apoelstra) [inline bip-tombriar-compressed-transactions.mediawiki:88]: `inputs` should be `input's` or `input`. Also you inconsistently capitalize Input in this section.

File: bip-tombriar-compressed-transactions.mediawiki [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+{| class="wikitable" style="margin:auto"
+|-
+! Name !! Width !! Description
+|-
+| Sequence || 0-5 Bytes || An Optional VarInt containing the sequence if it was non-standard.
+|-
+| Txid Blockheight || 1-5 Bytes || A VarInt Either containing 0 if this an uncompressed input, or it contains the offset from Minimum Blockheight for this Txid.
+|-
+| Txid/Signature Data || 65+ Bytes || Txid/Signatures are determined to be uncompressed either by the output script of the previous transaction, or if the Txid Blockheight is zero. For each Compressed Txid/Signature see [[#compressed-data|here]]. For each Uncompressed Txid/Signature see [[#uncompressed-data|here]].
```
Comment (apoelstra) [inline bip-tombriar-compressed-transactions.mediawiki:136]: When linking you should use the title of the section rather than `here`. This helps screen readers and other accessibility tools, and also helps sight-readers know where the link is going to go.

File: bip-tombriar-compressed-transactions.mediawiki [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+| Output Script || 2+ Bytes || A VLP-Bytestream containing the output script.
+|-
+| Amount || 1-9 Bytes || A VarInt containing the output amount.
+|}
+
+==== Ideal Transaction ====
+
+The target transaction for the most optimal compression was chosen
+based off the most common transactions that are likely to be used
+for purposes that requires the best compression.
```
Comment (apoelstra) [inline bip-tombriar-compressed-transactions.mediawiki:183]: I'd reword this because it's hard to tell whether you're about to describe an "ideal" transaction or a "common" one or what. I suggest

```
The compression scheme was designed to be optimal for a "typical" transaction spending a few close-in-age inputs and having one or two outputs. Here are size values for such a transaction which demonstrate the effectiveness of the compression.
```

File: bip-tombriar-compressed-transactions.mediawiki [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+The four main methods to achieve a lower transactions size are:
+
+1. Packing transaction metadata before it and each of its inputs and outputs to determine the structure of the data that follows.
+
+2. Replacing 32-bit numeric values with either variable-length integers (VarInts) or compact-integers (CompactSizes).
+
+3. Using compressed signatures and public key recovery upon decompression.
+
+4. Replacing the 36-byte txid/vout pair with a block height and index.
```
Comment (apoelstra) [inline bip-tombriar-compressed-transactions.mediawiki:225]: Both the "Motivation" and "Rationale" sections should go up top, after the abstract :). Nobody is going to read this far without a motivation.

File: bip-tombriar-compressed-transactions.mediawiki [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+|-
+| Input Hash Type || 0x00 for Taproot, 0x01 for Legacy || 7 Bits For Each Input
+|-
+| Output Script || Non-custom Scripts || 2 - 5 Bytes For Each Output
+|-
+| Output Amount || No Restrictions || (-1) - 7 Bytes For Each Output
+|}
+
+=== Motivation ===
+Bitcoin transactions currently contain a large amount of white space and padding. Specific fields are often one of a minimal number of possibilities. Leveraging these facts, we can create an encoding for 90% of Bitcoin transactions that are roughly 25-50% smaller.
```
Comment (apoelstra) [inline bip-tombriar-compressed-transactions.mediawiki:213]: I'd say "typical Bitcoin transactions" rather than just "Bitcoin transactions". Technically, no part of a transaction is "whitespace" or meaningless. It's just that *usually* the version, locktime, sequence numbers, etc., are one of a few values.

File: bip-tombriar-compressed-transactions.mediawiki [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+==== P2PKH ====
+
+===== Uncompressed =====
+<code>02000000015f5be26862482fe2fcc900f06ef26ee256fb205bc4773e5a402d0c1b88b82043000000006a473044022031a20f5d9212023b510599c9d53d082f8e07faaa2d51482e078f8e398cb50d770220635abd99220ad713a081c4f20b83cb3f491ed8bd032cb151a3521ed144164d9c0121027977f1b6357cead2df0a0a19570088a1eb9115468b2dfa01439493807d8f1294fdffffff0188130000000000001600142da377ed4978fefa043a58489912f8e28e16226200000000</code>
+
+===== Compressed =====
+<code>16b1ec7c2db1ec7d981431a20f5d9212023b510599c9d53d082f8e07faaa2d51482e078f8e398cb50d77635abd99220ad713a081c4f20b83cb3f491ed8bd032cb151a3521ed144164d9c8efefefe7d2da377ed4978fefa043a58489912f8e28e162262a608</code>
+
+== Acknowledgements ==
+Thank you to Andrew Poelstra who helped invent and develop the ideas in the proposal and the code in the reference implementation!
```
Comment (apoelstra) [inline bip-tombriar-compressed-transactions.mediawiki:268]: I appreciate the enthusiasm but I think this'd read better with a period instead of the exclamation point.

File: bip-tombriar-compressed-transactions.mediawiki (hunk 2 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki:84]: It's used for all script types but has special implications for P2TR and P2SH-WPKH. I reworded for clarity.

File: bip-tombriar-compressed-transactions.mediawiki (hunk 1 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki]: Fixed, Thanks!

File: bip-tombriar-compressed-transactions.mediawiki [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+| VLP-Bytestream || 2+ Bytes || A VarInt Length Prefixed Bytestream. Has a VarInt prefixed to determine the length.
+|}
+
+==== General Schema ====
+
+{| class="wikitable" style="margin:auto"
+|-
+! Name !! Width !! Description
+|-
+| Transaction metadata || 1 bytes || Information on the structure of the transaction. See [[#transaction-metadata|Transaction Metadata]]
```
[the remark this answers is not in the document]
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki]: Fixed, Thanks!

File: bip-tombriar-compressed-transactions.mediawiki (hunk 3 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki]: Fixed, Thanks!

File: bip-tombriar-compressed-transactions.mediawiki (hunk 4 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki]: Fixed, Thanks!

File: bip-tombriar-compressed-transactions.mediawiki (hunk 5 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki]: Fixed, Thanks!

File: bip-tombriar-compressed-transactions.mediawiki (hunk 6 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki]: Fixed, Thanks!

File: bip-tombriar-compressed-transactions.mediawiki (hunk 7 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki]: Fixed, Thanks!

File: bip-tombriar-compressed-transactions.mediawiki (hunk 8 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki]: Fixed, Thanks!

File: bip-tombriar-compressed-transactions.mediawiki [hunk 10]
```diff
@@ -0,0 +1,293 @@
+<pre>
+  BIP: ?
+  Layer: API/RPC
+  Title: Compressed Transactions RPC
+  Author: Tom Briar <tombriar11@protonmail.com>
+  Comments-URI: https://github.com/bitcoin/bitcoin/pull/29134
+                https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-August/021924.html
```
Comment (murchandamus) [inline bip-tombriar-compressed-transactions.mediawiki:7]: I think this is mixing up the "Comments-URI" and the "Post-History" headers.
```suggestion
  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-tombriar-compressed-transactions
```

File: bip-tombriar-compressed-transactions.mediawiki [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+  Status: Draft
+  Type: Standards Track
+  Created: 2024-02-01
+  License: BSD-3-Clause
```
Comment (murchandamus) [inline bip-tombriar-compressed-transactions.mediawiki:11]:
```suggestion
  License: BSD-3-Clause
  Post-History: https://github.com/bitcoin/bitcoin/pull/29134
                https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-August/021924.html
```

File: bip-tombriar-compressed-transactions.mediawiki [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+
+== Introduction ==
+
+=== Abstract ===
+This document proposes a serialization scheme for compressing Bitcoin transactions. The compressed Bitcoin transactions can reach a serialized size of less than 50% of the original serialized transaction. One method for compressing involves reducing the transaction outpoints in a potentially lossy way. Therefore, it is an optional path for compression. Compressing the outpoints is necessary for compressed transactions to reach less than 70% of the original size.
+
+=== Motivation ===
+Typical Bitcoin transactions usually contain a large amount of white space and padding due to specific fields that are often one of a minimal number of possibilities. We can use this fact and a few similar methods to create an encoding for 90% of Bitcoin transactions that are roughly 25-50% smaller.
+
+=== Rational ===
```
Comment (murchandamus) [inline bip-tombriar-compressed-transactions.mediawiki:22]:
```suggestion
=== Rationale ===
```

**Review by murchandamus — CHANGES_REQUESTED** [review]
Light first review. Looks pretty good, could you please add the mandatory "Backwards Compatibility" section (even if it just says that there are no concerns) and take a look at the preamble?

File: bip-tombriar-compressed-transactions.mediawiki (hunk 10 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki:8]: Fixed, Thanks!

File: bip-tombriar-compressed-transactions.mediawiki (hunk 12 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki]: Fixed, Thanks!

**Comment by TomBriar** [comment]
Thank you @murchandamus, all changes were accepted!

**Comment by murchandamus** [comment]
Thanks, I aim to give this a proper review soon

File: bip-tombriar-compressed-transactions.mediawiki [hunk 13]
```diff
@@ -0,0 +1,301 @@
+<pre>
+  BIP: ?
+  Layer: API/RPC
+  Title: Compressed Transactions RPC
```
Comment (murchandamus) [inline bip-tombriar-compressed-transactions.mediawiki:4]: This document seems to be focused on describing the scheme for compressing transactions and only mentions the RPC in context of the reference implementation. Perhaps it would make sense to drop the "RPC" from the title.
```suggestion
  Title: Compressed Transactions
```

File: bip-tombriar-compressed-transactions.mediawiki [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+  Author: Tom Briar <tombriar11@protonmail.com>
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-tombriar-compressed-transactions
+  Comments-URI: https://github.com/bitcoin/bitcoin/pull/29134
+                https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-August/021924.html
```
[the remark this answers is not in the document]
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki:8]: Fixed, Thanks!

File: bip-tombriar-compressed-transactions.mediawiki (hunk 13 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki:4]: Fixed, Thanks!

**Comment by murchandamus** [comment]
Hey @TomBriar, I’m not sure what happened here, it looks like you started working on this, but maybe you did not push the changes yet?

If the answer to my two open questions would be that alternative designs are not relevant, and the proposal is currently not implemented in any software that would be satisfactory answers to me—I just thought if those answers were different it would be nice to mention it.

As soon as you push your changes, I can take it from there. The only things missing would be to assign a number and merge it, which I can take care of.

**Comment by TomBriar** [comment]
@murchandamus Hi, Yes there is actually a relevant implementation I am finishing the finial touches on it today and I'll link it all together and have this ready to go by EOD. Thank you for your patience!

Thanks-
Tom.

**Comment by murchandamus** [comment]
I see, thanks for the update!

**Comment by TomBriar** [comment]
Hi,

Just pushed changes that address all of the above reviews, As stated, there is a project that I am aware of that will enable people to use steganography to encode data into images although it is not complete and I was not given an ETA. That said I have built a tool that implements the scheme described here. The hope is that with this BIP, future projects will all use this scheme for compressed Bitcoin transactions, and potentially bitcoin core itself can use parts of it in P2P or Block Storage.

https://github.com/TomBriar/compressed_transactions

Thanks-
Tom.

File: bip-tombriar-compressed-transactions.mediawiki [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+
+== Introduction ==
+
+=== Abstract ===
+This document proposes a serialization scheme for compressing Bitcoin transactions. The compressed Bitcoin transactions can reach a serialized size of less than 50% of the original serialized transaction. One method for compressing involves reducing the transaction outpoints in a potentially lossy way. Therefore, it is an optional path for compression. Compressing the outpoints is necessary for compressed transactions to reach less than 70% of the original size.
+
+=== Motivation ===
+Typical Bitcoin transactions usually contain a large amount of white space and padding due to specific fields that are often one of a minimal number of possibilities. We can use this fact and a few similar methods to create an encoding for 90% of Bitcoin transactions that are roughly 25-50% smaller.
+
+=== Rationale ===
```
[the remark this answers is not in the document]
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki:25]: https://github.com/TomBriar/compressed_transactions is a full implementation in rust

**Comment by murchandamus** [comment]
Hey, you wrote that you pushed, but I don’t see a change. Could you check whether the push went through?

File: bip-tombriar-compressed-transactions.mediawiki [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+
+== Introduction ==
+
+=== Abstract ===
+This document proposes a serialization scheme for compressing Bitcoin transactions. The compressed Bitcoin transactions can reach a serialized size of less than 50% of the original serialized transaction. One method for compressing involves reducing the transaction outpoints in a potentially lossy way. Therefore, it is an optional path for compression. Compressing the outpoints is necessary for compressed transactions to reach less than 70% of the original size.
+
+=== Motivation ===
+Typical Bitcoin transactions usually contain a large amount of white space and padding due to specific fields that are often one of a minimal number of possibilities. We can use this fact and a few similar methods to create an encoding for 90% of Bitcoin transactions that are roughly 25-50% smaller.
+
+There exists a working-in-progress app that allows the use of steganography to encode data in images to be passed around via various social media groups. When used in conjunction with this compression scheme and an elligator squared encryption, would allow for a very secure and private form of broadcasting bitcoin transactions.
```
Comment (murchandamus) [inline bip-tombriar-compressed-transactions.mediawiki:23]:
```suggestion
There exists a working-in-progress app that allows the use of steganography to encode data in images to be passed around via various social media groups. When used in conjunction with this compression scheme and an elligator squared encryption, this would allow for a very secure and private form of broadcasting bitcoin transactions.
```

**Review by murchandamus — APPROVED** [review]
Spotted a missing word here, but otherwise LGTM.

**Comment by murchandamus** [comment]
I assign this BIP 337

**Comment by TomBriar** [comment]
Hi, Thanks!

I made that change and fixed the BIP number field.

-Tom

File: bip-tombriar-compressed-transactions.mediawiki (hunk 16 above)
Comment (TomBriar) [reply bip-tombriar-compressed-transactions.mediawiki]: Fixed, Thanks!

**Comment by murchandamus** [comment]
I fixed the Comments-URI header, added an entry in the README.mediawiki table, and moved your BIP to the numbered file

**Comment by TomBriar** [comment]
Awesome thank you!

**Comment by murchandamus** [comment]
Thank you, and congrats! :)
